lib/connections/service.go

```diff
@@ -440,8 +466,49 @@ func (s *service) dialDevices(ctx context.Context, now time.Time, cfg config.Con
 		}
 
 		dialTargets := s.resolveDialTargets(ctx, now, cfg, deviceCfg, nextDialAt, initial, priorityCutoff)
-		if conn, ok := s.dialParallel(ctx, deviceCfg.DeviceID, dialTargets); ok {
+		if len(dialTargets) > 0 {
+			queue = append(queue, dialQueueEntry{
+				id:       deviceCfg.DeviceID,
+				lastSeen: stats[deviceCfg.DeviceID.String()].LastSeen,
+				targets:  dialTargets,
+			})
+		}
+	}
+
+	// Sort the queue with the most recently seen device at the head,
+	// increasing the likelihood of connecting to a device that we're
+	// already almost up to date with, index wise.
+	sort.Slice(queue, func(a, b int) bool {
+		return queue[a].lastSeen.After(queue[b].lastSeen)
+	})
+
+	// Shuffle the part of the connection queue that are devices we haven't
+	// connected to recently, so that if we only try a limited set of
+	// devices (or they in turn have limits and we're trying to load balance
+	// over several) and the usual ones are down it won't be the same ones
+	// in the same order every time.
+	idx := 0
+	cutoff := time.Now().Add(-recentlySeenCutoff)
+	for idx < len(queue) {
+		if queue[idx].lastSeen.Before(cutoff) {
+			break
+		}
+		idx++
+	}
+	if idx < len(queue)-1 {
+		rand.Shuffle(queue[idx:])
+	}
+
+	// Perform dials according to the queue, stopping when we've reached the
+	// allowed additional number of connections (if limited).
```

Comment:
There's a bug here in that we might establish a connection here only to have it dropped in a few milliseconds due to the other side having paused us or reached a limit.

We can't easily predict the future here so one idea that strikes me is that we should remember a connection being dropped in the handshake and penalise that device in future connection attempts (sorting it at the back of the queue).

Ideally we should send `Close` messages with a cause when devices are paused and such, so we can know this on the receiving side and reduce connection attempts towards that device in general...

Reply:
Sending a close message definitely is a good idea.

The general problem of counting soon to fail connections could also be solved by sending a result channel together with the connection on `s.conns`, on which the handle-loop sends a bool whether the connection has been accepted/successful or not.

Reply:
I was thinking of adding a statistic for connection life time, tracked the same way we track "last seen", updated in Closed(). When sorting devices for connection purposes we can then take into account the last connection life time, or average of the last 5 connection life times, or something like that. If it's less than a second we can conclude the device has issues talking to us and try them less often.

I don't think the channel thing will work with the current setup, as on the dialing side the connection looks accepted and  goes all the way through model.AddConnection() and such. Then it gets torn down some short interval later, probably while we're preparing or sending cluster configs.